[SYCL] Get rid of ifunc GNU extension uses

smaslov-intel opened the pull request:
Change resolution of PI interfaces from load-time to run-time. This is a step towards real plugins system, but more importantly it gets rid of "ifunc" GNU extension to be able to build on Windows. The key changes are in the definitions of "dispatch" in pi.hpp and pi_opencl.cpp

Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

keryell commented on sycl/include/CL/sycl/detail/pi.hpp:
```diff
@@ -34,136 +43,70 @@ namespace pi {
+  // TODO: implement a more mature and controllable tracing of PI calls.
+  class PiCall {
+    PiResult  m_Result;
+    static bool   m_TraceEnabled;
```
Weird spacing on these 2 lines

smaslov-intel replied:
Fixed

keryell commented on sycl/source/detail/pi.cpp:
```diff
@@ -35,6 +28,25 @@ bool piUseBackend(PiBackend Backend) {
 
+// Definitions of the PI dispatch entries, they will be initialized
+// at their first use with piInitialize.
+#define _PI_API(api) decltype(::api) * api = nullptr;
```
Is it not scary to have a global `api` variable?

smaslov-intel replied:
It will need to be globals in the long run. The PI plugins (shared libraries) are going to be discovered on the system by locating these globals, and they will connect with SYCL RT (and possibly other tools) with these globals. Remember that these globals are pointers to extern ""C" PI API.

keryell replied:
Yes but then why such a trivial name? It seems like a very high conflict probablility...
It could be `_sycl_backend_pi_api_` or whatever.

keryell commented on sycl/source/detail/pi_opencl.cpp:
```diff
@@ -79,26 +76,21 @@ pi_result OCL(piextDeviceSelectBinary)(
-decltype(ocl_api) OCL(pi_api) __attribute__((ifunc ("__resolve_" #pi_api)));
+// Forward calls to OpenCL RT.
+#define _PI_CL(pi_api, ocl_api)   \
+const decltype(::pi_api) * pi_api##OclPtr = \
```
Another global variable, `pi_api`?

smaslov-intel replied:
This one is temporary. We could change the OpenCL PI plugin to be hooked like any other plugin (in run-time).

keryell commented on sycl/source/detail/pi.cpp:
```diff
@@ -35,6 +28,25 @@ bool piUseBackend(PiBackend Backend) {
+    piDie("Unknown SYCL_BE");
+  }
+  #define _PI_API(api) \
+    extern const decltype(::api) * api##OclPtr; \
```
Probably nicer to align the `\` vertically for macro definitions like it is done is the code around

smaslov-intel replied:
Maybe

keryell commented on sycl/source/detail/pi_opencl.cpp:
```diff
@@ -79,26 +76,21 @@ pi_result OCL(piextDeviceSelectBinary)(
-}                                           \
-decltype(ocl_api) OCL(pi_api) __attribute__((ifunc ("__resolve_" #pi_api)));
+// Forward calls to OpenCL RT.
+#define _PI_CL(pi_api, ocl_api)   \
```
Do not change `\` alignment style

smaslov-intel replied:
OK

v-klochkov commented:
This patch needs re-base - I see some conflicts right now when try using cherry-pick of this commit.

smaslov-intel commented:
Rebased and addressed comments.

> I do not understand the relation with the PR name.

The key changes are into pi* dispatch to not use GNU "ifunc" extension for call forwarding.
